File: IterativePostOrder1Stack.cpp
#include<iostream>
#include<stack>
#include<vector>
using namespace std;

class Solution{
    public:
        vector<int> postorderTraversal(TreeNode* root){
            vector<int> postorder;
            stack<TreeNode*> s1;
            TreeNode* current = root;
            
            while(current != NULL || s1.empty()){
                if(current != NULL){
                    s1.push(current);
                    current= current->left;
                }

                else{
                    TreeNode* temp = s1.top()->right;
                    if(temp == NULL){
                        temp=s1.top();
                        s1.pop();
                        postorder.push_back(temp->val);
                        while(!s1.empty() && temp == s1.top()->right){
                            temp = s1.top();
                            s1.pop();
                            postorder.push_back(temp->val);
                        }
                    }
                    else{
                        current = temp;
                    }
                }
            }
            return postorder;
        }
};

File: IterativePostOrder1Stack.java
import java.util.*;
import java.io.*;

class Solution{
    public List<Integer> postorderTraversal(TreeNode root){
        Stack<TreeNode> s1 = new Stack<TreeNode>();
        List<Integer> postOrder = new ArrayList<Integer>();

        if(root == null)
            return postOrder;

        TreeNode current = root;
        while(current != null || !s1.isEmpty()){
            if(current != null){
                s1.push(current);
                current = current.left;
            }
            else{
                TreeNode temp = s1.peek().right;
                if(temp == null){
                    temp = s1.pop();
                    postOrder.add(temp.val);
                }
                while( !s1.isEmpty() && temp == s1.peek().right){
                    temp = s1.pop();
                    postOrder.add(temp.val);
                }
                    
            }
            else{
                current =temp;
            }
        }
    }
    return postOrder;
}